Snakefile

```diff
@@ -1,6 +1,154 @@
+SAMPLES, = glob_wildcards(config["input_path"]+"/{sample}.bam")
+
+# probably need to specify here (or in config) the species or the location of the genome/annotations
+
+rule all:
+    input: expand(["out/{sample}/{sample}.fastq.val", "out/{sample}.txt"], sample=SAMPLES), "done.txt"
+
+
+rule check_bam:
+    input:
+        config["input_path"]+ "/{sample}.bam"
+    output:
+        bam_check = "out/{sample}/{sample}_1.bam_checked"
+    conda:
+        "envs/samtools.yml"
+    shell:
+        """
+        samtools quickcheck {input}
+	    if [ $? -ne 0 ]; then
+		    echo "ERROR: {input} is not a valid BAM file" >&2
+	    else
+		    touch {output.bam_check}
+        fi
+        """
 
 rule bam_to_fastq:
+    input:
+        bam = config["input_path"]+"/{sample}.bam",
+        check_bam = rules.check_bam.output.bam_check
+    output:
+        fastq = "out/{sample}/{sample}.fq.gz",
+    conda: 
+        "envs/samtools.yml"
+    threads: 4
+    shell:
+        """
+        samtools fastq --threads {threads} -0 /dev/null {input.bam} > {output.fastq}
+        """
+
+# here if else statement could be modified to run iRAP/ISL for PE and SE data
+checkpoint validating_fastq:
+    input:
+        fastq = rules.bam_to_fastq.output.fastq,
+    output:
+        val_fastq = "out/{sample}/{sample}.fastq.val",
+    params:
+        fastq = "out/{sample}/{sample}.fastq.gz",
+    conda:
+        "envs/fastq_utils.yml"
+    shell:
+        """
+        fastq_info {input.fastq}  
+        if [ $? -ne 0 ]; then
+            #rm -rf {input.fastq}
+            echo "ERROR: Failed fastq validation {input.fastq}"
+            mv {input.fastq} {params.fastq}
+        else
+            echo "validation successful"
+        fi
+
+        touch {output.val_fastq}
+        """
+
+def aggregate_fastq(wildcards):
+    checkpoint_output = checkpoints.validating_fastq.get(**wildcards).output[0]
+    print(checkoint_output)
+    return expand("out/{sample}/{fq}_fastqc.html",
+        sample=wildcards.sample,
+        fq=glob_wildcards(os.path.join(checkpoint_output, "{fq}.fastq.gz")).fq)
+
+rule qc:
+    input:
+        check = rules.validating_fastq.output.val_fastq
+    output:
+        fqc = "out/{sample}/{fq}_1_fastqc.html",
+    params:
+        fq = "out/{sample}/{fq}.fastq.gz",
+        fqc_dir = "out/{sample}"
+    conda:
+        "envs/fastqc.yml"
+    threads: 4
+    shell:
+        """
+        fastqc -c {threads} {params.fq} --outdir={params.fqc_dir}
+        """
+
+rule run_irap:
+    input:
+        fastq=rules.bam_to_fastq.output.fastq,
+    output: "out/{sample}.txt"
+	conda: "envs/isl.yaml"
+    params:
+        private_script=config["private_script"],
+        conf=config["irap_config"],
+        strand="both",
+        irapMem=4096000000,
+        irapDataOption="",
+		filename="{sample}"
+	resources: mem_mb=10000	
+	shell:
+        """
+        source {params.private_script}/gtex_bulk_env.sh
+        source {params.private_script}/gtex_bulk_init.sh 
```

Comment:
Line not needed.  `gtex_bulk_init.sh` should be run only once to set up the dir tree (outside of Snakemake), before any libraries are processed.

Reply:
This can be moved a separate rule if needed, or leave it as it is as its not harming or recreating dir structure..